src/System/TimedScopes/TimedScope.cs

```diff
@@ -0,0 +1,983 @@
+		/// <param name="statusCode">status code to set success value from</param>
+		/// <remarks>500 or above is considered failure,
+		/// all other values considered success. However, if the success value is already set,
+		/// calling this method does not override the success value.
+		/// The status code is stored on the scope in any case.</remarks>
+		public void SetSuccessFromStatusCode(HttpStatusCode statusCode)
+		{
+			if (!IsSuccessful.HasValue)
+			{
+				IsSuccessful = (uint)statusCode < (uint)HttpStatusCode.InternalServerError;
+			}
+
+			if (string.IsNullOrWhiteSpace(TimedScopeData.Data(TimedScopeDataKeys.InternalOnly.StatusCode)))
+			{
+				AddLoggingValue(TimedScopeDataKeys.InternalOnly.StatusCode, statusCode.ToString());
+			}
+		}
+
+
+		/// <summary>
+		/// Is the timed scope disposed
+		/// </summary>
+		private bool IsDisposed { get; set; }
+
+
+		/// <summary>
+		/// End the timed scope explicitly
+		/// </summary>
+		private void EndScope(IMachineInformation machineInformation)
+		{
+			if (Interlocked.CompareExchange(ref m_isScopeEnded, 1, 0) == 0)
+			{
+				EndTick = Stopwatch.GetTimestamp();
+
+				PerfDiagnostics?.Stop();
+
+				LogEnd(machineInformation);
+
+				IsScopeActive = false;
+			}
+		}
+
+
+		/// <summary>
+		/// Logs the scope end to ULS
+		/// </summary>
+		private void LogEnd(IMachineInformation machineInformation)
+		{
+			if (!IsSuccessful.HasValue)
+			{
+				ULSLogging.LogTraceTag(0x2385029d /* tag_97qk3 */, Categories.TimingGeneral, Levels.Warning,
+					"Result not set for scope {0}. Considered as SystemError", Name);
+
+				Result = TimedScopeResult.SystemError;
+				FailureDescription = InternalFailureDescription.UnknownResultAsSystemError;
+			}
+
+			CorrelationData scopeData = ConstructCorrelationDataEntries(machineInformation);
+			ScopeLogger.LogScopeEnd(this, scopeData);
+
+			ReplayEventConfigurator.ConfigureReplayEventsOnScopeEnd(this);
+		}
+
+		/// <summary>
+		/// Dispose of the timed scope
+		/// </summary>
+		public void Dispose()
+		{
+			Dispose(true, MachineInformation);
+			GC.SuppressFinalize(this);
+		}
+
+
+		/// <summary>
+		/// Dispose of the timed scope
+		/// </summary>
+		/// <param name="disposing">Should dispose or not</param>
+		/// <param name="machineInformation">Machine Information</param>
+		protected virtual void Dispose(bool disposing, IMachineInformation machineInformation)
+		{
+			if (disposing && !IsDisposed)
+			{
+				IsDisposed = true;
+				EndScope(machineInformation);
+			}
+		}
+
+
+		/// <summary>
+		/// Timed scope data, added to the log output
+		/// </summary>
+		private CorrelationData TimedScopeData { get; set; }
+
+
+		/// <summary>
+		/// Constructs the timed scope correlation data
+		/// </summary>
+		/// <returns>Correlation data</returns>
+		private CorrelationData ConstructCorrelationDataEntries(IMachineInformation machineInformation)
+		{
+
+			// TODO what is the difference between CrossCutting.Corelation.CurrentCorelation and TimedScopeData
```

Comment:
> TODO wh [](start = 5, length = 8)

please address all TODOs, the ones that can't be addressed in this PR, please create GitHub issues

Reply:
This one was temporary added by me (I missed it during the cleanup), I remove it. Concerning other TODOs, what do you mean by saying 'address them'?

Reply:
The issue concerning the timed scope logging: [https://github.com/microsoft/Omex/issues/81](https://github.com/microsoft/Omex/issues/81)